// @tanstack/ai-event-client/dist/esm/devtools-middleware.js
import { aiEventClient } from "./index.js";
function buildEventContext(ctx) {
  return {
    requestId: ctx.requestId,
    streamId: ctx.streamId,
    provider: ctx.provider,
    model: ctx.model,
    clientId: ctx.conversationId,
    source: ctx.source,
    systemPrompts: ctx.systemPrompts.length > 0 ? ctx.systemPrompts : void 0,
    toolNames: ctx.toolNames,
    options: ctx.options,
    modelOptions: ctx.modelOptions,
    messageCount: ctx.messageCount,
    hasTools: ctx.hasTools,
    streaming: ctx.streaming
  };
}
function getContentString(content) {
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";
  return content.map(
    (part) => part && typeof part === "object" && part.type === "text" ? String(part.content ?? "") : ""
  ).join("") || "";
}
function devtoolsMiddleware() {
  let localMessageId = null;
  let localAccumulatedContent = "";
  let currentIteration = -1;
  let iterationStartTime = 0;
  const activeToolCalls = /* @__PURE__ */ new Map();
  return {
    name: "devtools",
    onStart(ctx) {
      aiEventClient.emit("text:request:started", {
        ...buildEventContext(ctx),
        timestamp: Date.now()
      });
      const messages = ctx.messages;
      const messagesToEmit = ctx.conversationId ? messages.slice(-1).filter((m) => m.role === "user") : messages;
      messagesToEmit.forEach((message, index) => {
        const messageIndex = ctx.conversationId ? messages.length - 1 : index;
        const messageId = ctx.createId("msg");
        const base = buildEventContext(ctx);
        const content = getContentString(message.content);
        aiEventClient.emit("text:message:created", {
          ...base,
          messageId,
          role: message.role,
          content,
          toolCalls: message.toolCalls,
          messageIndex,
          timestamp: Date.now()
        });
        if (message.role === "user") {
          aiEventClient.emit("text:message:user", {
            ...base,
            messageId,
            role: "user",
            content,
            messageIndex,
            timestamp: Date.now()
          });
        }
      });
    },
    onIteration(ctx, info) {
      const now = Date.now();
      if (currentIteration >= 0) {
        aiEventClient.emit("text:iteration:completed", {
          ...buildEventContext(ctx),
          iteration: currentIteration,
          messageId: localMessageId || void 0,
          duration: now - iterationStartTime,
          finishReason: "tool_calls",
          timestamp: now
        });
      }
      currentIteration = info.iteration;
      iterationStartTime = now;
      localMessageId = info.messageId;
      localAccumulatedContent = "";
      aiEventClient.emit("text:iteration:started", {
        ...buildEventContext(ctx),
        iteration: info.iteration,
        messageId: info.messageId,
        timestamp: now
      });
      aiEventClient.emit("text:message:created", {
        ...buildEventContext(ctx),
        messageId: info.messageId,
        role: "assistant",
        content: "",
        timestamp: now
      });
    },
    onChunk(ctx, chunk) {
      const base = buildEventContext(ctx);
      switch (chunk.type) {
        case "TEXT_MESSAGE_CONTENT": {
          if (chunk.content) {
            localAccumulatedContent = chunk.content;
          } else {
            localAccumulatedContent += chunk.delta;
          }
          aiEventClient.emit("text:chunk:content", {
            ...base,
            messageId: localMessageId || void 0,
            content: localAccumulatedContent,
            delta: chunk.delta,
            timestamp: Date.now()
          });
          break;
        }
        case "TOOL_CALL_START": {
          const toolIndex = chunk.index ?? 0;
          activeToolCalls.set(chunk.toolCallId, {
            toolName: chunk.toolName,
            index: toolIndex
          });
          aiEventClient.emit("text:chunk:tool-call", {
            ...base,
            messageId: localMessageId || void 0,
            toolCallId: chunk.toolCallId,
            toolName: chunk.toolName,
            index: toolIndex,
            arguments: "",
            timestamp: Date.now()
          });
          break;
        }
        case "TOOL_CALL_ARGS": {
          const active = activeToolCalls.get(chunk.toolCallId);
          aiEventClient.emit("text:chunk:tool-call", {
            ...base,
            messageId: localMessageId || void 0,
            toolCallId: chunk.toolCallId,
            toolName: active?.toolName ?? "",
            index: active?.index ?? 0,
            arguments: chunk.delta,
            timestamp: Date.now()
          });
          break;
        }
        case "TOOL_CALL_END": {
          activeToolCalls.delete(chunk.toolCallId);
          aiEventClient.emit("text:chunk:tool-result", {
            ...base,
            messageId: localMessageId || void 0,
            toolCallId: chunk.toolCallId,
            result: chunk.result || "",
            timestamp: Date.now()
          });
          break;
        }
        case "RUN_FINISHED": {
          aiEventClient.emit("text:chunk:done", {
            ...base,
            messageId: localMessageId || void 0,
            finishReason: chunk.finishReason,
            usage: chunk.usage,
            timestamp: Date.now()
          });
          if (chunk.usage) {
            aiEventClient.emit("text:usage", {
              ...base,
              messageId: localMessageId || void 0,
              usage: chunk.usage,
              timestamp: Date.now()
            });
          }
          break;
        }
        case "RUN_ERROR": {
          const err = chunk.error;
          aiEventClient.emit("text:chunk:error", {
            ...base,
            messageId: localMessageId || void 0,
            error: err?.message ?? String(chunk.error),
            timestamp: Date.now()
          });
          break;
        }
        case "STEP_FINISHED": {
          if (chunk.content || chunk.delta) {
            aiEventClient.emit("text:chunk:thinking", {
              ...base,
              messageId: localMessageId || void 0,
              content: chunk.content || "",
              delta: chunk.delta,
              timestamp: Date.now()
            });
          }
          break;
        }
      }
    },
    onToolPhaseComplete(ctx, info) {
      const base = buildEventContext(ctx);
      if (info.toolCalls.length > 0) {
        aiEventClient.emit("text:message:created", {
          ...base,
          messageId: localMessageId ?? ctx.createId("msg"),
          role: "assistant",
          content: localAccumulatedContent || "",
          toolCalls: info.toolCalls,
          timestamp: Date.now()
        });
      }
      for (const approval of info.needsApproval) {
        aiEventClient.emit("tools:approval:requested", {
          ...base,
          messageId: localMessageId || void 0,
          toolCallId: approval.toolCallId,
          toolName: approval.toolName,
          input: approval.input,
          approvalId: approval.approvalId,
          timestamp: Date.now()
        });
      }
      for (const clientTool of info.needsClientExecution) {
        aiEventClient.emit("tools:input:available", {
          ...base,
          messageId: localMessageId || void 0,
          toolCallId: clientTool.toolCallId,
          toolName: clientTool.toolName,
          input: clientTool.input,
          timestamp: Date.now()
        });
      }
      for (const result of info.results) {
        aiEventClient.emit("tools:call:completed", {
          ...base,
          messageId: localMessageId || void 0,
          toolCallId: result.toolCallId,
          toolName: result.toolName,
          result: result.result,
          duration: result.duration ?? 0,
          timestamp: Date.now()
        });
        const content = JSON.stringify(result.result);
        aiEventClient.emit("text:message:created", {
          ...base,
          messageId: ctx.createId("msg"),
          role: "tool",
          content,
          timestamp: Date.now()
        });
      }
    },
    onFinish(ctx, info) {
      const now = Date.now();
      if (currentIteration >= 0) {
        aiEventClient.emit("text:iteration:completed", {
          ...buildEventContext(ctx),
          iteration: currentIteration,
          messageId: localMessageId || void 0,
          duration: now - iterationStartTime,
          finishReason: info.finishReason || void 0,
          usage: info.usage,
          timestamp: now
        });
      }
      aiEventClient.emit("text:request:completed", {
        ...buildEventContext(ctx),
        content: info.content,
        messageId: localMessageId || void 0,
        finishReason: info.finishReason || void 0,
        usage: info.usage,
        duration: info.duration,
        timestamp: now
      });
    }
  };
}
export {
  devtoolsMiddleware
};
