// @tanstack/ai-client/dist/esm/connection-adapters.js
function mergeHeaders(customHeaders) {
  if (!customHeaders) {
    return {};
  }
  if (customHeaders instanceof Headers) {
    const result = {};
    customHeaders.forEach((value, key) => {
      result[key] = value;
    });
    return result;
  }
  return customHeaders;
}
async function* readStreamLines(reader, abortSignal) {
  try {
    const decoder = new TextDecoder();
    let buffer = "";
    while (true) {
      if (abortSignal?.aborted) {
        break;
      }
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";
      for (const line of lines) {
        if (line.trim()) {
          yield line;
        }
      }
    }
    if (buffer.trim()) {
      yield buffer;
    }
  } finally {
    reader.releaseLock();
  }
}
function normalizeConnectionAdapter(connection) {
  if (!connection) {
    throw new Error("Connection adapter is required");
  }
  const hasConnect = "connect" in connection;
  const hasSubscribe = "subscribe" in connection;
  const hasSend = "send" in connection;
  if (hasConnect && (hasSubscribe || hasSend)) {
    throw new Error(
      "Connection adapter must provide either connect or both subscribe and send, not both modes"
    );
  }
  if (hasSubscribe && hasSend) {
    return {
      subscribe: connection.subscribe.bind(connection),
      send: connection.send.bind(connection)
    };
  }
  if (!hasConnect) {
    throw new Error(
      "Connection adapter must provide either connect or both subscribe and send"
    );
  }
  let activeBuffer = [];
  let activeWaiters = [];
  function push(chunk) {
    const waiter = activeWaiters.shift();
    if (waiter) {
      waiter(chunk);
    } else {
      activeBuffer.push(chunk);
    }
  }
  return {
    subscribe(abortSignal) {
      const myBuffer = activeBuffer.splice(0);
      const myWaiters = [];
      activeBuffer = myBuffer;
      activeWaiters = myWaiters;
      return (async function* () {
        while (!abortSignal?.aborted) {
          let chunk;
          if (myBuffer.length > 0) {
            chunk = myBuffer.shift();
          } else {
            chunk = await new Promise((resolve) => {
              const onAbort = () => resolve(null);
              myWaiters.push((c) => {
                abortSignal?.removeEventListener("abort", onAbort);
                resolve(c);
              });
              abortSignal?.addEventListener("abort", onAbort, { once: true });
            });
          }
          if (chunk !== null) yield chunk;
        }
      })();
    },
    async send(messages, data, abortSignal) {
      let hasTerminalEvent = false;
      try {
        const stream2 = connection.connect(messages, data, abortSignal);
        for await (const chunk of stream2) {
          if (chunk.type === "RUN_FINISHED" || chunk.type === "RUN_ERROR") {
            hasTerminalEvent = true;
          }
          push(chunk);
        }
        if (!abortSignal?.aborted && !hasTerminalEvent) {
          push({
            type: "RUN_FINISHED",
            runId: `run-${Date.now()}`,
            model: "connect-wrapper",
            timestamp: Date.now(),
            finishReason: "stop"
          });
        }
      } catch (err) {
        if (!abortSignal?.aborted && !hasTerminalEvent) {
          push({
            type: "RUN_ERROR",
            timestamp: Date.now(),
            error: {
              message: err instanceof Error ? err.message : "Unknown error in connect()"
            }
          });
        }
        throw err;
      }
    }
  };
}
function fetchServerSentEvents(url, options = {}) {
  return {
    async *connect(messages, data, abortSignal) {
      const resolvedUrl = typeof url === "function" ? url() : url;
      const resolvedOptions = typeof options === "function" ? await options() : options;
      const requestHeaders = {
        "Content-Type": "application/json",
        ...mergeHeaders(resolvedOptions.headers)
      };
      const requestBody = {
        messages,
        data,
        ...resolvedOptions.body
      };
      const fetchClient = resolvedOptions.fetchClient ?? fetch;
      const response = await fetchClient(resolvedUrl, {
        method: "POST",
        headers: requestHeaders,
        body: JSON.stringify(requestBody),
        credentials: resolvedOptions.credentials || "same-origin",
        signal: abortSignal || resolvedOptions.signal
      });
      if (!response.ok) {
        throw new Error(
          `HTTP error! status: ${response.status} ${response.statusText}`
        );
      }
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error("Response body is not readable");
      }
      for await (const line of readStreamLines(reader, abortSignal)) {
        const data2 = line.startsWith("data: ") ? line.slice(6) : line;
        if (data2 === "[DONE]") continue;
        try {
          const parsed = JSON.parse(data2);
          yield parsed;
        } catch (parseError) {
          console.warn("Failed to parse SSE chunk:", data2);
        }
      }
    }
  };
}
function fetchHttpStream(url, options = {}) {
  return {
    async *connect(messages, data, abortSignal) {
      const resolvedUrl = typeof url === "function" ? url() : url;
      const resolvedOptions = typeof options === "function" ? await options() : options;
      const requestHeaders = {
        "Content-Type": "application/json",
        ...mergeHeaders(resolvedOptions.headers)
      };
      const requestBody = {
        messages,
        data,
        ...resolvedOptions.body
      };
      const fetchClient = resolvedOptions.fetchClient ?? fetch;
      const response = await fetchClient(resolvedUrl, {
        method: "POST",
        headers: requestHeaders,
        body: JSON.stringify(requestBody),
        credentials: resolvedOptions.credentials || "same-origin",
        signal: abortSignal || resolvedOptions.signal
      });
      if (!response.ok) {
        throw new Error(
          `HTTP error! status: ${response.status} ${response.statusText}`
        );
      }
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error("Response body is not readable");
      }
      for await (const line of readStreamLines(reader, abortSignal)) {
        try {
          const parsed = JSON.parse(line);
          yield parsed;
        } catch (parseError) {
          console.warn("Failed to parse HTTP stream chunk:", line);
        }
      }
    }
  };
}
function stream(streamFactory) {
  return {
    async *connect(messages, data) {
      yield* streamFactory(messages, data);
    }
  };
}
function rpcStream(rpcCall) {
  return {
    async *connect(messages, data) {
      yield* rpcCall(messages, data);
    }
  };
}
export {
  fetchHttpStream,
  fetchServerSentEvents,
  normalizeConnectionAdapter,
  rpcStream,
  stream
};
