// @tanstack/ai-client/dist/esm/generation-client.js
import { GENERATION_EVENTS } from "./generation-types.js";
import { parseSSEResponse } from "./sse-parser.js";
class GenerationClient {
  constructor(options) {
    this.result = null;
    this.isLoading = false;
    this.error = void 0;
    this.status = "idle";
    this.abortController = null;
    this.connection = options.connection;
    this.fetcher = options.fetcher;
    this.body = options.body ?? {};
    this.callbacksRef = {
      onResult: options.onResult,
      onError: options.onError,
      onProgress: options.onProgress,
      onChunk: options.onChunk,
      onResultChange: options.onResultChange,
      onLoadingChange: options.onLoadingChange,
      onErrorChange: options.onErrorChange,
      onStatusChange: options.onStatusChange
    };
  }
  /**
   * Trigger a generation request.
   * Only one generation can be in-flight at a time; calling generate()
   * while already generating will be a no-op.
   */
  async generate(input) {
    if (this.isLoading) return;
    this.setIsLoading(true);
    this.setStatus("generating");
    this.setError(void 0);
    const abortController = new AbortController();
    this.abortController = abortController;
    const { signal } = abortController;
    try {
      if (this.fetcher) {
        const result = await this.fetcher(input, { signal });
        if (signal.aborted) return;
        if (result instanceof Response) {
          await this.processStream(parseSSEResponse(result, signal));
        } else {
          this.setResult(result);
          this.setStatus("success");
        }
      } else if (this.connection) {
        const mergedData = { ...this.body, ...input };
        const stream = this.connection.connect([], mergedData, signal);
        await this.processStream(stream);
      } else {
        throw new Error(
          "GenerationClient requires either a connection or fetcher option"
        );
      }
    } catch (err) {
      if (signal.aborted) return;
      const error = err instanceof Error ? err : new Error(String(err));
      this.setError(error);
      this.setStatus("error");
      this.callbacksRef.onError?.(error);
    } finally {
      this.abortController = null;
      this.setIsLoading(false);
    }
  }
  /**
   * Process a stream of AG-UI events from the streaming connection adapter.
   */
  async processStream(source) {
    for await (const chunk of source) {
      if (this.abortController?.signal.aborted) break;
      this.callbacksRef.onChunk?.(chunk);
      switch (chunk.type) {
        case "CUSTOM": {
          if (chunk.name === GENERATION_EVENTS.RESULT) {
            this.setResult(chunk.value);
          } else if (chunk.name === GENERATION_EVENTS.PROGRESS) {
            const { progress, message } = chunk.value;
            this.callbacksRef.onProgress?.(progress, message);
          }
          break;
        }
        case "RUN_FINISHED": {
          this.setStatus("success");
          break;
        }
        case "RUN_ERROR": {
          throw new Error(chunk.error.message);
        }
      }
    }
  }
  /**
   * Abort any in-flight generation request.
   */
  stop() {
    if (this.abortController) {
      this.abortController.abort();
      this.abortController = null;
    }
    this.setIsLoading(false);
    if (this.status === "generating") {
      this.setStatus("idle");
    }
  }
  /**
   * Clear the result, error, and return to idle state.
   */
  reset() {
    this.stop();
    this.setResult(null);
    this.setError(void 0);
    this.setStatus("idle");
  }
  /**
   * Update options without recreating the client.
   */
  updateOptions(options) {
    if (options.body !== void 0) {
      this.body = options.body ?? {};
    }
    if (options.onResult !== void 0) {
      this.callbacksRef.onResult = options.onResult;
    }
    if (options.onError !== void 0) {
      this.callbacksRef.onError = options.onError;
    }
    if (options.onProgress !== void 0) {
      this.callbacksRef.onProgress = options.onProgress;
    }
    if (options.onChunk !== void 0) {
      this.callbacksRef.onChunk = options.onChunk;
    }
  }
  // ===========================
  // Getters
  // ===========================
  getResult() {
    return this.result;
  }
  getIsLoading() {
    return this.isLoading;
  }
  getError() {
    return this.error;
  }
  getStatus() {
    return this.status;
  }
  // ===========================
  // Private state setters
  // ===========================
  setResult(rawResult) {
    if (rawResult === null) {
      this.result = null;
      this.callbacksRef.onResultChange?.(null);
      return;
    }
    if (this.callbacksRef.onResult) {
      const transformed = this.callbacksRef.onResult(rawResult);
      if (transformed === null) {
        return;
      }
      if (transformed !== void 0) {
        this.result = transformed;
        this.callbacksRef.onResultChange?.(this.result);
        return;
      }
    }
    this.result = rawResult;
    this.callbacksRef.onResultChange?.(this.result);
  }
  setIsLoading(isLoading) {
    this.isLoading = isLoading;
    this.callbacksRef.onLoadingChange?.(isLoading);
  }
  setError(error) {
    this.error = error;
    this.callbacksRef.onErrorChange?.(error);
  }
  setStatus(status) {
    this.status = status;
    this.callbacksRef.onStatusChange?.(status);
  }
}
export {
  GenerationClient
};
