// @tanstack/ai-client/dist/esm/video-generation-client.js
import { GENERATION_EVENTS } from "./generation-types.js";
import { parseSSEResponse } from "./sse-parser.js";
class VideoGenerationClient {
  constructor(options) {
    this.result = null;
    this.jobId = null;
    this.videoStatus = null;
    this.isLoading = false;
    this.error = void 0;
    this.status = "idle";
    this.abortController = null;
    this.connection = options.connection;
    this.fetcher = options.fetcher;
    this.body = options.body ?? {};
    this.callbacksRef = {
      onResult: options.onResult,
      onError: options.onError,
      onProgress: options.onProgress,
      onChunk: options.onChunk,
      onJobCreated: options.onJobCreated,
      onStatusUpdate: options.onStatusUpdate,
      onResultChange: options.onResultChange,
      onLoadingChange: options.onLoadingChange,
      onErrorChange: options.onErrorChange,
      onStatusChange: options.onStatusChange,
      onJobIdChange: options.onJobIdChange,
      onVideoStatusChange: options.onVideoStatusChange
    };
  }
  /**
   * Trigger video generation.
   * Only one generation can be in-flight at a time.
   */
  async generate(input) {
    if (this.isLoading) return;
    this.setIsLoading(true);
    this.setStatus("generating");
    this.setError(void 0);
    this.setJobId(null);
    this.setVideoStatus(null);
    const abortController = new AbortController();
    this.abortController = abortController;
    const { signal } = abortController;
    try {
      if (this.fetcher) {
        await this.generateWithFetcher(input, signal);
      } else if (this.connection) {
        const mergedData = { ...this.body, ...input };
        const stream = this.connection.connect([], mergedData, signal);
        await this.processStream(stream);
      } else {
        throw new Error(
          "VideoGenerationClient requires either a connection or fetcher option"
        );
      }
    } catch (err) {
      if (signal.aborted) return;
      const error = err instanceof Error ? err : new Error(String(err));
      this.setError(error);
      this.setStatus("error");
      this.callbacksRef.onError?.(error);
    } finally {
      this.abortController = null;
      this.setIsLoading(false);
    }
  }
  /**
   * Direct fetcher mode: call fetcher and set result.
   */
  async generateWithFetcher(input, signal) {
    if (!this.fetcher) return;
    const result = await this.fetcher(input, { signal });
    if (signal.aborted) return;
    if (result instanceof Response) {
      await this.processStream(parseSSEResponse(result, signal));
    } else {
      this.setResult(result);
      this.setStatus("success");
    }
  }
  /**
   * Process a stream of AG-UI events from the streaming connection adapter.
   * The server handles the polling loop and streams status updates.
   */
  async processStream(source) {
    for await (const chunk of source) {
      if (this.abortController?.signal.aborted) break;
      this.callbacksRef.onChunk?.(chunk);
      switch (chunk.type) {
        case "CUSTOM": {
          if (chunk.name === GENERATION_EVENTS.VIDEO_JOB_CREATED) {
            const { jobId } = chunk.value;
            this.setJobId(jobId);
            this.callbacksRef.onJobCreated?.(jobId);
          } else if (chunk.name === GENERATION_EVENTS.VIDEO_STATUS) {
            const statusInfo = chunk.value;
            this.setVideoStatus(statusInfo);
            this.callbacksRef.onStatusUpdate?.(statusInfo);
            if (statusInfo.progress !== void 0) {
              this.callbacksRef.onProgress?.(statusInfo.progress);
            }
          } else if (chunk.name === GENERATION_EVENTS.RESULT) {
            this.setResult(chunk.value);
          } else if (chunk.name === GENERATION_EVENTS.PROGRESS) {
            const { progress, message } = chunk.value;
            this.callbacksRef.onProgress?.(progress, message);
          }
          break;
        }
        case "RUN_FINISHED": {
          this.setStatus("success");
          break;
        }
        case "RUN_ERROR": {
          throw new Error(chunk.error.message);
        }
      }
    }
  }
  /**
   * Abort any in-flight generation or polling.
   */
  stop() {
    if (this.abortController) {
      this.abortController.abort();
      this.abortController = null;
    }
    this.setIsLoading(false);
    if (this.status === "generating") {
      this.setStatus("idle");
    }
  }
  /**
   * Clear all state and return to idle.
   */
  reset() {
    this.stop();
    this.setResult(null);
    this.setJobId(null);
    this.setVideoStatus(null);
    this.setError(void 0);
    this.setStatus("idle");
  }
  /**
   * Update options without recreating the client.
   */
  updateOptions(options) {
    if (options.body !== void 0) {
      this.body = options.body ?? {};
    }
    if (options.onResult !== void 0) {
      this.callbacksRef.onResult = options.onResult;
    }
    if (options.onError !== void 0) {
      this.callbacksRef.onError = options.onError;
    }
    if (options.onProgress !== void 0) {
      this.callbacksRef.onProgress = options.onProgress;
    }
    if (options.onChunk !== void 0) {
      this.callbacksRef.onChunk = options.onChunk;
    }
    if (options.onJobCreated !== void 0) {
      this.callbacksRef.onJobCreated = options.onJobCreated;
    }
    if (options.onStatusUpdate !== void 0) {
      this.callbacksRef.onStatusUpdate = options.onStatusUpdate;
    }
  }
  // ===========================
  // Getters
  // ===========================
  getResult() {
    return this.result;
  }
  getJobId() {
    return this.jobId;
  }
  getVideoStatus() {
    return this.videoStatus;
  }
  getIsLoading() {
    return this.isLoading;
  }
  getError() {
    return this.error;
  }
  getStatus() {
    return this.status;
  }
  // ===========================
  // Private state setters
  // ===========================
  setResult(rawResult) {
    if (rawResult === null) {
      this.result = null;
      this.callbacksRef.onResultChange?.(null);
      return;
    }
    if (this.callbacksRef.onResult) {
      const transformed = this.callbacksRef.onResult(rawResult);
      if (transformed === null) {
        return;
      }
      if (transformed !== void 0) {
        this.result = transformed;
        this.callbacksRef.onResultChange?.(this.result);
        return;
      }
    }
    this.result = rawResult;
    this.callbacksRef.onResultChange?.(this.result);
  }
  setJobId(jobId) {
    this.jobId = jobId;
    this.callbacksRef.onJobIdChange?.(jobId);
  }
  setVideoStatus(status) {
    this.videoStatus = status;
    this.callbacksRef.onVideoStatusChange?.(status);
  }
  setIsLoading(isLoading) {
    this.isLoading = isLoading;
    this.callbacksRef.onLoadingChange?.(isLoading);
  }
  setError(error) {
    this.error = error;
    this.callbacksRef.onErrorChange?.(error);
  }
  setStatus(status) {
    this.status = status;
    this.callbacksRef.onStatusChange?.(status);
  }
}
export {
  VideoGenerationClient
};
