// @tanstack/ai-react/dist/esm/use-generate-video.js
import { VideoGenerationClient } from "@tanstack/ai-client";
import { useId, useState, useRef, useMemo, useEffect, useCallback } from "react";
function useGenerateVideo(options) {
  const hookId = useId();
  const clientId = options.id || hookId;
  const [result, setResult] = useState(null);
  const [jobId, setJobId] = useState(null);
  const [videoStatus, setVideoStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(void 0);
  const [status, setStatus] = useState("idle");
  const optionsRef = useRef(options);
  optionsRef.current = options;
  const client = useMemo(() => {
    const opts = optionsRef.current;
    const baseOptions = {
      id: clientId,
      body: opts.body,
      onResult: (r) => optionsRef.current.onResult?.(r),
      onError: (e) => optionsRef.current.onError?.(e),
      onProgress: (p, m) => optionsRef.current.onProgress?.(p, m),
      onChunk: (c) => optionsRef.current.onChunk?.(c),
      onJobCreated: (id) => optionsRef.current.onJobCreated?.(id),
      onStatusUpdate: (s) => optionsRef.current.onStatusUpdate?.(s),
      onResultChange: setResult,
      onLoadingChange: setIsLoading,
      onErrorChange: setError,
      onStatusChange: setStatus,
      onJobIdChange: setJobId,
      onVideoStatusChange: setVideoStatus
    };
    if (opts.connection) {
      return new VideoGenerationClient({
        ...baseOptions,
        connection: opts.connection
      });
    }
    if (opts.fetcher) {
      return new VideoGenerationClient({
        ...baseOptions,
        fetcher: opts.fetcher
      });
    }
    throw new Error(
      "useGenerateVideo requires either a connection or fetcher option"
    );
  }, [clientId]);
  useEffect(() => {
    client.updateOptions({ body: options.body });
  }, [client, options.body]);
  useEffect(() => {
    return () => {
      client.stop();
    };
  }, [client]);
  const generate = useCallback(
    async (input) => {
      await client.generate(input);
    },
    [client]
  );
  const stop = useCallback(() => {
    client.stop();
  }, [client]);
  const reset = useCallback(() => {
    client.reset();
  }, [client]);
  return {
    generate,
    result,
    jobId,
    videoStatus,
    isLoading,
    error,
    status,
    stop,
    reset
  };
}
export {
  useGenerateVideo
};
