// @tanstack/ai-react/dist/esm/use-realtime-chat.js
import { useState, useRef, useEffect, useCallback } from "react";
import { RealtimeClient } from "@tanstack/ai-client";
const emptyFrequencyData = new Uint8Array(128);
const emptyTimeDomainData = new Uint8Array(128).fill(128);
function useRealtimeChat(options) {
  const [status, setStatus] = useState("idle");
  const [mode, setMode] = useState("idle");
  const [messages, setMessages] = useState([]);
  const [pendingUserTranscript, setPendingUserTranscript] = useState(null);
  const [pendingAssistantTranscript, setPendingAssistantTranscript] = useState(null);
  const [error, setError] = useState(null);
  const [inputLevel, setInputLevel] = useState(0);
  const [outputLevel, setOutputLevel] = useState(0);
  const [vadMode, setVADModeState] = useState(
    options.vadMode ?? "server"
  );
  const clientRef = useRef(null);
  const optionsRef = useRef(options);
  optionsRef.current = options;
  const animationFrameRef = useRef(null);
  if (!clientRef.current) {
    clientRef.current = new RealtimeClient({
      getToken: optionsRef.current.getToken,
      adapter: optionsRef.current.adapter,
      tools: optionsRef.current.tools,
      instructions: optionsRef.current.instructions,
      voice: optionsRef.current.voice,
      autoPlayback: optionsRef.current.autoPlayback,
      autoCapture: optionsRef.current.autoCapture,
      vadMode: optionsRef.current.vadMode,
      outputModalities: optionsRef.current.outputModalities,
      temperature: optionsRef.current.temperature,
      maxOutputTokens: optionsRef.current.maxOutputTokens,
      semanticEagerness: optionsRef.current.semanticEagerness,
      onStatusChange: (newStatus) => {
        setStatus(newStatus);
      },
      onModeChange: (newMode) => {
        setMode(newMode);
        optionsRef.current.onModeChange?.(newMode);
      },
      onMessage: (message) => {
        setMessages((prev) => [...prev, message]);
        optionsRef.current.onMessage?.(message);
      },
      onError: (err) => {
        setError(err);
        optionsRef.current.onError?.(err);
      },
      onConnect: () => {
        setError(null);
        optionsRef.current.onConnect?.();
      },
      onDisconnect: () => {
        optionsRef.current.onDisconnect?.();
      },
      onInterrupted: () => {
        setPendingAssistantTranscript(null);
        optionsRef.current.onInterrupted?.();
      }
    });
    clientRef.current.onStateChange((state) => {
      setPendingUserTranscript(state.pendingUserTranscript);
      setPendingAssistantTranscript(state.pendingAssistantTranscript);
    });
  }
  const client = clientRef.current;
  useEffect(() => {
    function updateLevels() {
      if (clientRef.current?.audio) {
        setInputLevel(clientRef.current.audio.inputLevel);
        setOutputLevel(clientRef.current.audio.outputLevel);
      }
      animationFrameRef.current = requestAnimationFrame(updateLevels);
    }
    if (status === "connected") {
      updateLevels();
    }
    return () => {
      if (animationFrameRef.current) {
        cancelAnimationFrame(animationFrameRef.current);
        animationFrameRef.current = null;
      }
    };
  }, [status]);
  useEffect(() => {
    return () => {
      clientRef.current?.destroy();
    };
  }, []);
  const connect = useCallback(async () => {
    setError(null);
    setMessages([]);
    setPendingUserTranscript(null);
    setPendingAssistantTranscript(null);
    await client.connect();
  }, [client]);
  const disconnect = useCallback(async () => {
    await client.disconnect();
  }, [client]);
  const startListening = useCallback(() => {
    client.startListening();
  }, [client]);
  const stopListening = useCallback(() => {
    client.stopListening();
  }, [client]);
  const interrupt = useCallback(() => {
    client.interrupt();
  }, [client]);
  const sendText = useCallback(
    (text) => {
      client.sendText(text);
    },
    [client]
  );
  const sendImage = useCallback(
    (imageData, mimeType) => {
      client.sendImage(imageData, mimeType);
    },
    [client]
  );
  const getInputFrequencyData = useCallback(() => {
    return clientRef.current?.audio?.getInputFrequencyData() ?? emptyFrequencyData;
  }, []);
  const getOutputFrequencyData = useCallback(() => {
    return clientRef.current?.audio?.getOutputFrequencyData() ?? emptyFrequencyData;
  }, []);
  const getInputTimeDomainData = useCallback(() => {
    return clientRef.current?.audio?.getInputTimeDomainData() ?? emptyTimeDomainData;
  }, []);
  const getOutputTimeDomainData = useCallback(() => {
    return clientRef.current?.audio?.getOutputTimeDomainData() ?? emptyTimeDomainData;
  }, []);
  const setVADMode = useCallback(
    (newMode) => {
      setVADModeState(newMode);
    },
    []
  );
  return {
    // Connection state
    status,
    error,
    connect,
    disconnect,
    // Conversation state
    mode,
    messages,
    pendingUserTranscript,
    pendingAssistantTranscript,
    // Voice control
    startListening,
    stopListening,
    interrupt,
    // Text input
    sendText,
    // Image input
    sendImage,
    // Audio visualization
    inputLevel,
    outputLevel,
    getInputFrequencyData,
    getOutputFrequencyData,
    getInputTimeDomainData,
    getOutputTimeDomainData,
    // VAD control
    vadMode,
    setVADMode
  };
}
export {
  useRealtimeChat
};
