// @tanstack/ai-openai/dist/esm/realtime/token.js
import { getOpenAIApiKeyFromEnv } from "../utils/client.js";
const OPENAI_REALTIME_SESSIONS_URL = "https://api.openai.com/v1/realtime/sessions";
function openaiRealtimeToken(options = {}) {
  const apiKey = getOpenAIApiKeyFromEnv();
  return {
    provider: "openai",
    async generateToken() {
      const model = options.model ?? "gpt-4o-realtime-preview";
      const response = await fetch(OPENAI_REALTIME_SESSIONS_URL, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json"
        },
        body: JSON.stringify({ model })
      });
      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(
          `OpenAI realtime session creation failed: ${response.status} ${errorText}`
        );
      }
      const sessionData = await response.json();
      return {
        provider: "openai",
        token: sessionData.client_secret.value,
        expiresAt: sessionData.client_secret.expires_at * 1e3,
        config: {
          model: sessionData.model
        }
      };
    }
  };
}
export {
  openaiRealtimeToken
};
