// @tanstack/ai-gemini/dist/esm/adapters/image.js
import { BaseImageAdapter } from "@tanstack/ai/adapters";
import { createGeminiClient, generateId, getGeminiApiKeyFromEnv } from "../utils/client.js";
import { validatePrompt, validateImageSize, validateNumberOfImages, parseNativeImageSize, sizeToAspectRatio } from "../image/image-provider-options.js";
class GeminiImageAdapter extends BaseImageAdapter {
  constructor(config, model) {
    super({}, model);
    this.kind = "image";
    this.name = "gemini";
    this.client = createGeminiClient(config);
  }
  async generateImages(options) {
    const { model, prompt } = options;
    validatePrompt({ prompt, model });
    if (this.isGeminiImageModel(model)) {
      return this.generateWithGeminiApi(options);
    }
    validateImageSize(model, options.size);
    validateNumberOfImages(model, options.numberOfImages);
    const config = this.buildImagenConfig(options);
    const response = await this.client.models.generateImages({
      model,
      prompt,
      config
    });
    return this.transformImagenResponse(model, response);
  }
  isGeminiImageModel(model) {
    return model.startsWith("gemini-");
  }
  async generateWithGeminiApi(options) {
    const { model, prompt, size, numberOfImages, modelOptions } = options;
    const parsedSize = size ? parseNativeImageSize(size) : void 0;
    const augmentedPrompt = numberOfImages && numberOfImages > 1 ? `${prompt} Generate ${numberOfImages} distinct images.` : prompt;
    const config = {
      // Include TEXT so the model can interleave descriptions between images
      responseModalities: ["TEXT", "IMAGE"],
      ...parsedSize && {
        imageConfig: {
          ...parsedSize.aspectRatio && {
            aspectRatio: parsedSize.aspectRatio
          },
          ...parsedSize.resolution && {
            imageSize: parsedSize.resolution
          }
        }
      },
      ...modelOptions
    };
    const response = await this.client.models.generateContent({
      model,
      contents: augmentedPrompt,
      config
    });
    return this.transformGeminiResponse(model, response);
  }
  transformGeminiResponse(model, response) {
    const images = [];
    const parts = response.candidates?.[0]?.content?.parts ?? [];
    for (const part of parts) {
      if (part.inlineData?.data && typeof part.inlineData.data === "string" && part.inlineData.data.length > 0) {
        images.push({ b64Json: part.inlineData.data });
      }
    }
    return {
      id: generateId(this.name),
      model,
      images,
      usage: void 0
    };
  }
  buildImagenConfig(options) {
    const { size, numberOfImages, modelOptions } = options;
    return {
      numberOfImages: numberOfImages ?? 1,
      // Map size to aspect ratio if provided (modelOptions.aspectRatio will override)
      aspectRatio: size ? sizeToAspectRatio(size) : void 0,
      ...modelOptions
    };
  }
  transformImagenResponse(model, response) {
    const images = (response.generatedImages ?? []).map(
      (item) => ({
        b64Json: item.image?.imageBytes,
        revisedPrompt: item.enhancedPrompt
      })
    );
    return {
      id: generateId(this.name),
      model,
      images,
      usage: void 0
    };
  }
}
function createGeminiImage(model, apiKey, config) {
  return new GeminiImageAdapter({ apiKey, ...config }, model);
}
function geminiImage(model, config) {
  const apiKey = getGeminiApiKeyFromEnv();
  return createGeminiImage(model, apiKey, config);
}
export {
  GeminiImageAdapter,
  createGeminiImage,
  geminiImage
};
